File: 22_generate_parentheses.rb
# @param {Integer} n
# @return {String[]}
def generate_parenthesis(n)
  result = []
  
  result = generate("", 0, 0, n, result)
end

def generate(string, left, right, n, result)
  if string.length == 2*n
    result << string
  end
  if left < n
    generate(string + "(", left+1, right, n, result)
  end
  if right < left
    generate(string + ")", left, right+1, n, result)
  end
    
  result
end

File: 22_generate_parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
  var result = [];

  var generate = function (string, left, right) {
    if (string.length === 2 * n) {
      result.push(string);
      return;
    }
    if (left < n) {
      generate(string + "(", left + 1, right);
    }
    if (right < left) {
      generate(string + ")", left, right + 1);
    }
  };

  generate("", 0, 0);
  return result;
};